drivers/hmis/app/graphql/types/hmis_schema/enrollment.rb

```diff
@@ -27,6 +27,26 @@ def self.configuration
       Hmis::Hud::Enrollment.hmis_configuration(version: '2024')
     end
 
+    # Override the "field" function to perform field-level authorization.
+    # If user lacks sufficient access, the field will be resolved as null.
+    #
+    # This is necessary because the user may have access to view
+    # "full" enrollment details for some Enrollments, but only "limited" access
+    # to other Enrollments.
+    def self.field(name, type = nil, **kwargs)
+      # See Types::BaseField `authorized?` function.
+      super(name, type, permission: :can_view_enrollment_details, **kwargs)
+    end
+
+    # Add a separate "field" function to use for summary fields which are visible
+    # to users with "can_view_limited_enrollment_details".
+    #
+    # No field-level authorization is needed here, because the enrollment necessarily
+    # has some level of visibility if it's being resolved.
+    def self.summary_field(name, type = nil, **kwargs)
+      field(name, type, **kwargs, permission: nil)
+    end
```

Comment:
This looks fine in terms of functionality. But I did want to float an idea for another approach here, particularly since this pattern might spread out into other resolvers. Feel free to ignore it if it seems like too much of a jump.

I know we have a pattern of creating `x_field` methods in our resolvers (such as `enrollments_field`, etc). There may be an opportunity to try a different approach here, particularly since this is a different dimension (authorization rather than data loading). 

The primary concern I have with the `x_field` pattern is that it creates a level of indirection that may be hard to understand. For example it may not be obvious that `assessments_field` will call `field` which we are overriding in this type, as opposed to passing it in explicitly with `assessments_field(permissions: xyz)`

I understand the current pattern addresses some important needs: preventing incorrect or missing permissions and also keeping our resolvers DRY. But maybe we could solve that in another way. One approach would be to change `field()` so that permissions are a required arg. Then we could use something like `with_options` to dry up the permissions declaration. Something like this:

```ruby
    # override field so permissions is required kwarg
    def self.field(name, type = nil, permissions:, **kwargs, &block)
      # See Types::BaseField `authorized?` function.
      super(name, type, permissions: permissions, **kwargs, &block)
    end

    # summary fields
    with_options permissions: [:can_view_enrollment_summary] do |type|
      type.field :household_id, ID, null: false
      type.assessments_field
      # etc
    end

    # detail fields
    with_options permissions: [:can_view_enrollment_details] do |type|
      type.field :household_id, ID, null: false
      # etc
    end

    # runtime error
    #  field :project, Types::HmisSchema::Project, null: false
    # this is okay
    #  field :project, Types::HmisSchema::Project, null: false, permissions: [:can_view_enrollment_details]


```

Reply:
Nice, I like that approach. I agree that the indirection with overriding `field` is confusing, it's not particularly a pattern I'd like to spread to other resolvers.

I'm not _quite_ following how `type.assessments_field` would work in your example, but I like the direction. I'm not a huge fan of the current `HasAssessments` / `assessments_field` pattern.

I think for the sake of getting this out by end of week, I need to merge this and get it up for QA. But I think your suggestion would be great and I captured it here https://www.pivotaltracker.com/story/show/186384085! I don't have a strong sense of whether this type of field authorization will be needed in other places. There are 2 other instances I have in mind: (1) redacting SSN/DOB on the Client if user lacks access, and (2) redacting fields on File if its confidential and user lacks access. Those are both done in their own way, not using the ruby-graphql field authorization mechanism. They are much more contained than this Enrollment situation though.